AppCenter/AppCenter/Internals/Model/Util/MSAppCenterUserDefaults.m

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSAppCenterUserDefaults.h"
+#import "MSAppCenterInternal.h"
+#import "MSLogger.h"
+
+static NSString *const kMSAppCenterUserDefaultsMigratedKeyFormat = @"310%@UserDefaultsMigratedKey";
+
+static MSAppCenterUserDefaults *sharedInstance = nil;
+static dispatch_once_t onceToken;
+
+@implementation MSAppCenterUserDefaults
+
++ (instancetype)shared {
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[MSAppCenterUserDefaults alloc] init];
+  });
+  return sharedInstance;
+}
+
++ (void)resetSharedInstance {
+
+  // Reset the once_token so dispatch_once will run again.
+  onceToken = 0;
+  sharedInstance = nil;
+}
+
+- (void)migrateKeys:(NSDictionary *)migratedKeys forService:(NSString *)service {
+  NSNumber *serviceMigrated = [self objectForKey:[NSString stringWithFormat:kMSAppCenterUserDefaultsMigratedKeyFormat, service]];
+  if (serviceMigrated) {
+    return;
+  }
+  MSLogVerbose([MSAppCenter logTag], @"Migrating the old NSDefaults keys to new ones.");
+  for (NSObject *newKey in migratedKeys) {
+    NSAssert([newKey isKindOfClass:[NSString class]], @"Unsupported type");
+    id<NSObject> oldKey = migratedKeys[newKey];
+    NSString *newKeyString = (NSString *)newKey;
+    if ([oldKey isKindOfClass:[NSString class]]) {
+      id value = [[NSUserDefaults standardUserDefaults] objectForKey:(NSString *)oldKey];
+      [self swapKeys:(NSString *)oldKey newKey:newKeyString value:value];
+    } else {
+      NSAssert([oldKey isKindOfClass:[MSUserDefaultsPrefixKey class]], @"Unsupported type");
+
+      // List all the keys starting with oldKey.
+      NSString *oldKeyPrefix = ((MSUserDefaultsPrefixKey *)oldKey).keyPrefix;
+      NSArray *userDefaultKeys = [[[NSUserDefaults standardUserDefaults] dictionaryRepresentation] allKeys];
+      for (NSString *userDefaultsKey in userDefaultKeys) {
+        if ([userDefaultsKey hasPrefix:oldKeyPrefix]) {
+          NSString *suffix = [userDefaultsKey stringByReplacingOccurrencesOfString:oldKeyPrefix withString:@""];
+          NSString *newKeyWithSuffix = [newKeyString stringByAppendingString:suffix];
+          id value = [[NSUserDefaults standardUserDefaults] objectForKey:userDefaultsKey];
+          [self swapKeys:userDefaultsKey newKey:newKeyWithSuffix value:value];
+        }
+      }
+    }
+  }
+  [self setObject:@YES forKey:[NSString stringWithFormat:kMSAppCenterUserDefaultsMigratedKeyFormat, service]];
+}
+
+- (void)swapKeys:(NSString *)oldKey newKey:(NSString *)newKey value:(id)value {
+  if (value == nil) {
+    return;
+  }
+  [[NSUserDefaults standardUserDefaults] setObject:value forKey:newKey];
+  [[NSUserDefaults standardUserDefaults] removeObjectForKey:oldKey];
+  MSLogVerbose([MSAppCenter logTag], @"Migrating key %@ -> %@", oldKey, newKey);
+}
+
+- (NSString *)getAppCenterKeyFrom:(NSObject *)key {
```

Comment:
Why is key not `NSString *`? Same for `objectForKey:` , `setObject:forKey:`, `removeObjectForKey:`?

Reply:
Based on my feedback https://github.com/microsoft/appcenter-sdk-apple/pull/2013#discussion_r402982589

So, it's to support prefix rule